fix(lang): patch FAB's LocaleView to redirect to previous page

pomegranited opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
Overrides FAB's LocaleView to redirect to the request's referrer instead of relying on the session page history as [suggested here](https://github.com/apache/superset/issues/12768#issuecomment-2376141202).

This isn't a perfect solution, but it keeps the change very close to the actual issue.

Note that this view is marked out in the tests as an "unsecured" view, because we need to be able to change the current locale even when logged out (so that the login page can be presented in the user's preferred language).

**Rejected alternatives**

1. Call `update_redirect()` whenever a GET view is rendered ([like this](https://github.com/apache/superset/compare/master...open-craft:superset:jill/update_redirect)).
   This is more in line with how Flask AppBuilder works (and was [recommended here](https://github.com/dpgaspar/Flask-AppBuilder/pull/1679#discussion_r702709749)).
   However, this only works for pages rendered by Superset's backend, and most Superset pages are rendered by the frontend.
2. Hook into the [frontend's action logging](https://github.com/apache/superset/blob/7c903236498281f34970012e701250edef1b1493/superset-frontend/src/views/App.tsx#L54-L69) to update the page history, e.g when a `LOG_ACTIONS_SPA_NAVIGATION` event is logged to the backend.
  This approach felt even hackier than the one posted here, and farther from the issue it addresses.
  Also, Flask AppBuilder only provides the `update_redirect()` method which uses the current `request.url`, so we'd need to add another method that could use a different `url`.

Fixes: https://github.com/apache/superset/issues/12768
Part of: https://github.com/apache/superset/issues/30334
Fixes: https://github.com/openedx/openedx-aspects/issues/299

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:

https://github.com/user-attachments/assets/7bcdefa5-46d5-4173-98cc-884384af233d

After:

https://github.com/user-attachments/assets/70b32340-7f23-482d-a33b-8347aa5000b0

### TESTING INSTRUCTIONS

1. Load Superset with multiple languages enabled.
2. Visit any page.
3. Change your current language using the language switcher menu on the top right.
4. Ensure that your language is changed and you're redirect back to the page in step 2.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Fixes https://github.com/apache/superset/issues/12768
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

pomegranited committed:
fix: patch FAB's LocaleView

to redirect to the request referrer instead of relying on the session
page_history.

This change is needed because most of Superset's page views are handled
by the frontend, and as such, not stored in the session page_history.

pomegranited committed:
fix: ensure the referrer is safe before redirecting

pomegranited commented on superset/initialization/__init__.py:
```diff
@@ -701,3 +701,21 @@ class SupersetIndexView(IndexView):
+        served by the frontend (and thus not added to the session's page_history
+        stack).
+        """
+        if locale not in self.appbuilder.bm.languages:
+            abort(404, description="Locale not supported.")
```
Leaving this one up to the reviewers - see original [LocaleView](https://github.com/dpgaspar/Flask-AppBuilder/blob/418ab8a93907669be4ccbb99d7aefa5283f3e013/flask_appbuilder/babel/views.py#L7-L19)

rusackas commented:
Installing the pre-commit hook should resolve at least some of the CI issues. 

/
Superset uses Git pre-commit hooks courtesy of [pre-commit](https://pre-commit.com/). To install run the following:
```
pip3 install -r requirements/development.txt
pre-commit install
A series of checks will now run when you make a git commit.
```
Alternatively it is possible to run pre-commit by running pre-commit manually:
```
pre-commit run --all-files
```

pomegranited committed:
fix: types and ruff format linting

pomegranited committed:
test: adds patch_flask_locale to list of allowed unsecured views

Because we need to be able to change the locale even when logged out,
this view cannot be protected.

pomegranited commented on superset/initialization/__init__.py:
```diff
@@ -701,3 +702,34 @@
+        redirect_to = request.headers.get("Referer")
+        if not redirect_to or not self.is_safe_url(redirect_to):
+            redirect_to = self.get_redirect()
+        return redirect(redirect_to)
```
Leaving this one up to the reviewers - see original [LocaleView](https://github.com/dpgaspar/Flask-AppBuilder/blob/418ab8a93907669be4ccbb99d7aefa5283f3e013/flask_appbuilder/babel/views.py#L7-L19)

dpgaspar replied:
we only redirect to the referer after checking with `self.is_safe_url` so I think this one is not correct by CodeQL.

pomegranited replied:
Ah.. that makes sense, thanks! See [3dc36a2](https://github.com/apache/superset/pull/31692/commits/3dc36a2e309e3af8812cacdc63e8e2698dec4c01).

pomegranited commented:
Hi @rusackas :) Is there anything more I can do here to get this PR approved?

rusackas commented:
I was hoping someone more FABtastic than me would come along to review this. If not @dpgaspar, perhaps @mistercrunch can take a look.

pomegranited commented:
Hi @jpchev , would you like to review this PR for me? Might help get it merged if there's some community testing/support :)

jpchev commented:
> Hi @jpchev , would you like to review this PR for me? Might help get it merged if there's some community testing/support :)

yes, I've tested the changes and can confirm I am not redirected to another page after switching language

dpgaspar reviewed (commented):
Looks good, would be great if you could add a couple of tests

pomegranited committed:
fix: satisfy codeql

pomegranited committed:
tests: adds tests for the patched flask locale view

pomegranited commented:
hi @dpgaspar , is there anything else you need from me before this can be merged?

pomegranited commented:
Hi @rusackas , I've applied @dpgaspar 's suggestions, but this has been needing another review/approval for a while now.. What can I do to get this merged? Would love to make the next release candidate.

pomegranited commented:
@mistercrunch Can you help me get this PR merged?

mistercrunch commented on superset/initialization/__init__.py:
```diff
@@ -701,3 +703,34 @@ class SupersetIndexView(IndexView):
         return redirect("/superset/welcome/")
+
+    @staticmethod
+    def is_safe_url(target: str) -> bool:
```
might belong in `superset/utils/urls.py` would be surprised if we don't already have a similar method there, grepping for `urlparse` or `next=` could point to a similar method.

mistercrunch replied:
otherwise PR LGTM

pomegranited replied:
Thank you @mistercrunch ! I didn't find a similar method anywhere, so [4bb4916](https://github.com/apache/superset/pull/31692/commits/4bb49161c04f4cedeaeba30d9fabf9889fd9497c) creates `superset.utils.urls.is_safe_redirect_url(source_url, target_url)`.

Does this look ok now?

pomegranited committed:
refactor: moves URL checking to utils.urls.is_safe_redirect_url

pomegranited commented on superset/initialization/__init__.py:
```diff
@@ -701,3 +703,34 @@
+        served by the frontend (and thus not added to the session's page_history
+        stack).
+        """
+        if locale not in self.appbuilder.bm.languages:
```
Mitigated by https://github.com/apache/superset/pull/31692/files#diff-ab057099067874ff6f10f77be2e3e6ed55ba08b7f743a40d5f3905fb836a7ee0R67

mistercrunch reviewed (approved):
LGTM - thank you for the contribution!

pomegranited commented:
Hi @rusackas , could this be merged soon?

mistercrunch commented on superset/utils/urls.py:
```diff
@@ -62,3 +62,20 @@ def is_secure_url(url: str) -> bool:
     return parsed_url.scheme == "https"
+
+
+def is_safe_redirect_url(source_url: str, target_url: str) -> bool:
```
sorry for the multi-phase review, but GPT recommends further checking here, and since this is security for XSS I'm thinking let's do it...

```
def is_safe_redirect_url(source_url: str, target_url: str) -> bool:
    if not target_url:
        return False
    joined = urljoin(source_url, target_url)  # resolves relative URLs
    parsed_source = urlparse(source_url)
    parsed_target = urlparse(joined)
    return (
        parsed_source.scheme == parsed_target.scheme and
        parsed_source.hostname == parsed_target.hostname
    )
```
GPT says: This handles edge cases like user-supplied target_url values starting with // (which browsers interpret as external redirects) or other relative path tricks. Using urljoin() ensures we're validating the fully resolved URL against the expected scheme and host. Safer for open internet exposure.

pomegranited replied:
Easy done, thank you for this security fix @mistercrunch ! Applied with [08010c9](https://github.com/apache/superset/pull/31692/commits/08010c9755ac5485a3480d615437d28cafacdcc7)

However, should we to use the joined `target_url` when doing the actual redirect?

pomegranited replied:
Ok, I've found what we needed in FAB: [get_safe_redirect](https://github.com/dpgaspar/Flask-AppBuilder/blob/6d44e6d581433dcea475764c4bb1270c24bbd6de/flask_appbuilder/utils/base.py#L36). My only worry is it's not exposed as an explicit part of the API, and so might change out from under us. But it's used in several places in the FAB code base, so it's a calculated risk.

rusackas commented:
@pomegranited just approved CI and it looks like this is mergeable pending feedback/action on the last comment from @mistercrunch.

rusackas reviewed (dismissed):
I was getting itchy over the merge button, so I'll block this awaiting feedback/implementation of the security-based code suggestion provided by @mistercrunch. CC @dpgaspar / @Antonio-RiveroMartnez for any other additional feedback or suggestions on that approach.

mistercrunch commented:
We just need to be super careful around the `?next={some_url}` pattern. Might be a good time to review what we already have here, and merge all codepaths that use this pattern into a state-of-the-art function that handles it all.

Couldn't find the pattern in the codebase after a quick search, maybe FAB exposes something we piggy-back on? @dpgaspar

pomegranited committed:
fix: resolve relative target URLs to avoid XSS attacks

mistercrunch reviewed (approved):
LGTM

mistercrunch commented:
Still would love for @dpgaspar to chime in, maybe FAB has builtin functions for that already

pomegranited committed:
refactor: use FAB's get_safe_redirect

pomegranited commented:
Hi @rusackas I found a FAB method to do what we need, so I think this resolves @mistercrunch 's concern?

dpgaspar commented:
> Hi @rusackas I found a FAB method to do what we need, so I think this resolves @mistercrunch 's concern?

Nice! yes `is_safe_redirect_url` or `get_safe_redirect`

dpgaspar reviewed (approved):
Nice, not a perfect solution but an improvement

pomegranited commented:
Hi @rusackas , I think we're good to merge now?

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/patch_flask_locale_redirect

pomegranited commented:
Hi @rusackas , I think I've addressed everything here, and we got an approval from our FAB expert.. can this be merged?

mistercrunch commented:
hitting a weird GH bug I've never seen before ->
<img width="886" alt="Screenshot 2025-04-15 at 9 34 10 AM" src="https://github.com/user-attachments/assets/c91af0d6-5c31-45af-8cfe-c4f498b7c695" />

Will try again a little later...

pomegranited commented:
Yay!! Thank you @mistercrunch @rusackas  @dpgaspar for your help here!
